File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap15/UDPClientWithTimeout.pl
#!/usr/bin/perl -w
use strict;

###  UDPClientWithTimeout.pl

use IO::Socket;                                                      #(A)
use constant MAX_MSG_LEN => 2048;                                    #(B)
use constant TIMEOUT => 5;
use constant MAX_RETRIES => 10;

die "usage: $0 host port" unless @ARGV == 2;                         #(C)

my $host = shift @ARGV;                                              #(D)
my $port = shift @ARGV;                                              #(E)

my $sock = IO::Socket::INET->new( Proto     => "udp",                #(F)
                                  PeerAddr  => $host,                #(G)
                                  PeerPort  => $port,                #(H)
                                )
                   or die $@;                                        #(I)
my $mesg_in;                                                         #(J)
my $retries = 1;                                                     #(K)

my $send_sequence_num = 1;
while ($retries < MAX_RETRIES ) {                                    #(L)
    $sock->send("$send_sequence_num: What's my lucky number today?")
                     or die "send() failed: $!";                     #(M)
    ++$send_sequence_num;
    eval {                                                           #(N)
        local $SIG{ALRM} = sub { ++$retries and die "timeout\n" };   #(O)
        alarm(TIMEOUT);                                              #(P)
        $sock->recv( $mesg_in, MAX_MSG_LEN )            
                            or die "recv() failed: $!";              #(Q)
        alarm(0);                                                    #(R)
        print "$mesg_in\n";                                          #(S)
    };
    if ( $@ eq "timeout\n" ) {                                       #(T)
        print "recv() timed out. Trying send() again " .
                    "with send sequence num: $send_sequence_num\n";  #(U)
    } else { last }                                                  #(V)
}

File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap15/UDPClientWithTimeout.py
#!/usr/bin/env python

###  UDPClientWithTimeout.py

import sys, socket, select, signal                                   #(A)

MAX_MSG_LEN = 2048                                                   #(B)
TIMEOUT = 5                                                          #(C)
MAX_RETRIES = 10                                                     #(D)

if len(sys.argv) != 3:                                            
    sys.exit( "Need exactly two command line arguments, the " +
              "first naming the host and the second the port" )  

host, port = sys.argv[1], int(sys.argv[2])                           #(E)

try:
    sock = socket.socket( socket.AF_INET, socket.SOCK_DGRAM )        #(F)
    sock.connect( (host, port) )                                     #(G)
except socket.error, (value, message):                               #(H)
    if sock:                                                         #(I)
        sock.close()                                                 #(J)
    print "Could not establish a client socket: " + message          #(K)
    sys.exit(1)                                                      #(L)

sock.settimeout( TIMEOUT )                                           #(M)
send_sequence_num = 1                                                #(N)
for retry in range( 1, MAX_RETRIES ):                                #(O)
    sock.send("%s: What's my lucky number today?"%send_sequence_num) #(P)
    send_sequence_num += 1                                           #(Q)
    mesg_in = None                                                   #(R)
    try:                                                             #(S)
        mesg_in = sock.recv( MAX_MSG_LEN )                           #(T)
    except socket.timeout, message:                                  #(U)
        print "recv() timed out. Trying send() again " + \
                    "with send sequence num: %s" % send_sequence_num #(V)
        continue                                                     #(W)
    print mesg_in                                                    #(X)
    break                                                            #(Y)
